// @microsoft/sp-property-pane/lib-commonjs/contentPanelController/ContentPanelController.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContentPanelController = void 0;
var tslib_1 = require("tslib");
var sp_core_library_1 = require("@microsoft/sp-core-library");
var sp_pages_content_panel_context_1 = require("@ms/sp-pages-content-panel-context");
var odsp_core_bundle_1 = require("@ms/odsp-core-bundle");
var sp_diagnostics_1 = require("@microsoft/sp-diagnostics");
var sp_loader_1 = require("@microsoft/sp-loader");
var propertyPaneConsumer_1 = require("../propertyPaneConsumer");
var Flights_1 = require("../common/Flights");
var constants_1 = require("../propertyPaneController/constants");
var excludedPanels_1 = require("./excludedPanels");
var includedWebpart_1 = require("./includedWebpart");
var logSource_1 = require("./logSource");
var KillSwitches_1 = require("../common/KillSwitches");
/**
 * @internal
 */
var ContentPanelController = /** @class */ (function () {
    function ContentPanelController() {
        var _this = this;
        /**
         * Dictionary of all the registered property pane consumers.
         */
        this._consumers = new Map();
        this.addOnUpdateListener = function (watcher) {
            return sp_pages_content_panel_context_1.SPPagesContentPanelProviderContext.getInstance().addOnUpdateListener(watcher);
        };
        /**
         * Register command listeners on mount.
         * Note this is called multiple times throughout this component's lifespan as a user navigates to different pages.
         */
        this._onContentPanelMount = function (_a) {
            var subscriberContext = _a.subscriberContext;
            sp_diagnostics_1._TraceLogger.logVerbose(logSource_1.logSource, 'ContentPanel mounted');
            var handlerId = 'ContentPanelController';
            subscriberContext.addCommandListener({
                handlerId: handlerId,
                commandId: sp_pages_content_panel_context_1.SPPagesContentPanelCommandId.propertyPaneConfigurationEvent,
                commandHandler: function (_a) {
                    var commandData = _a.commandData;
                    _this._onConfigurationEvent(commandData);
                }
            });
            subscriberContext.addCommandListener({
                handlerId: handlerId,
                commandId: sp_pages_content_panel_context_1.SPPagesContentPanelCommandId.propertyPaneFieldChange,
                commandHandler: function (_a) {
                    var consumerId = _a.consumerId, commandData = _a.commandData;
                    if (consumerId === _this.currentlyConfiguredConsumerId) {
                        _this._onPropertyPaneFieldChanged.apply(_this, commandData);
                    }
                }
            });
            subscriberContext.addCommandListener({
                handlerId: handlerId,
                commandId: sp_pages_content_panel_context_1.SPPagesContentPanelCommandId.refreshPropertyPane,
                commandHandler: function (_a) {
                    var consumerId = _a.consumerId;
                    if (_this.currentlyConfiguredConsumerId === consumerId && consumerId) {
                        _this._updateContentPanelData(_this.currentlyConfiguredConsumerId).catch(function (error) {
                            sp_diagnostics_1._TraceLogger.logError(logSource_1.logSource, error);
                        });
                    }
                }
            });
        };
        // method bindings
        this._onPropertyPaneFieldChanged = this._onPropertyPaneFieldChanged.bind(this);
        this._onConfigurationEvent = this._onConfigurationEvent.bind(this);
        this._fireConfigurationEvent = this._fireConfigurationEvent.bind(this);
        var qosMonitor = new sp_diagnostics_1._QosMonitor((0, sp_pages_content_panel_context_1.getQOSName)(logSource_1.COMPONENT_NAME, sp_pages_content_panel_context_1.SPPagesContentPanelQOSScenarios.loadChunk));
        sp_pages_content_panel_context_1.SPPagesContentPanelProviderContext.getInstance().addOnMountListener(function (context) {
            _this._onContentPanelMount(context);
            qosMonitor.writeSuccess();
        }, function (error) {
            var customError = new odsp_core_bundle_1.CustomError({
                message: (0, odsp_core_bundle_1.markPrivacyData)(String(error.message), 'safe'),
                innerError: error
            });
            sp_diagnostics_1._TraceLogger.logError(logSource_1.logSource, customError);
            qosMonitor.writeUnexpectedFailure('UnhandledError', customError);
        });
    }
    Object.defineProperty(ContentPanelController.prototype, "currentlyConfiguredConsumerId", {
        get: function () {
            return this._currentlyConfiguredConsumerId;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Determines if the Content Panel is supported in the current environment.
     */
    ContentPanelController.isContentPanelSupported = function () {
        var _a;
        return (!!((_a = sp_pages_content_panel_context_1.SPPagesContentPanelProviderContext.getInstance().subscriberContext) === null || _a === void 0 ? void 0 : _a.isPivotRendered(sp_pages_content_panel_context_1.SPPagesContentPanelPath.configurationTools)) && (0, Flights_1.isConfigurationToolsEnabled)());
    };
    ContentPanelController.prototype.registerOnPropertyPaneFieldChangedListener = function (instanceId, callback) {
        return function () { };
    };
    ContentPanelController.prototype.registerConsumer = function (instanceId, consumer) {
        sp_core_library_1.Validate.isNonemptyString(instanceId, 'instanceId');
        sp_core_library_1.Validate.isNotNullOrUndefined(consumer, 'consumer');
        this._consumers.set(instanceId, consumer);
        if (this._unregistredBackgroundRefreshConsumerId === instanceId) {
            this.requestAction(instanceId, 'BackgroundRefresh');
        }
    };
    ContentPanelController.prototype.isConsumerRegistered = function (instanceId) {
        return this._consumers.has(instanceId);
    };
    ContentPanelController.prototype.requestAction = function (toBeConfiguredConsumerId, propertyPaneAction, renderedByWebPart, context) {
        if (propertyPaneAction === void 0) { propertyPaneAction = 'Default'; }
        sp_diagnostics_1._TraceLogger.logVerbose(logSource_1.logSource, "request '".concat(propertyPaneAction, "' action for '").concat(toBeConfiguredConsumerId, "'. isOpen: ").concat(this.isOpen(), ", isAnyContentPanelOpen: ").concat(this.isAnyContentPanelOpen(), ", currentlyConfiguredConsumerId: ").concat(this.currentlyConfiguredConsumerId));
        this._unregistredBackgroundRefreshConsumerId = undefined;
        switch (propertyPaneAction) {
            case 'Open':
            case 'OpenDetails':
                this._openContentPanel(toBeConfiguredConsumerId, renderedByWebPart, propertyPaneAction === 'OpenDetails', context).catch(function (error) {
                    sp_diagnostics_1._TraceLogger.logError(logSource_1.logSource, error);
                });
                break;
            // Property pane should close.
            case 'Close':
            case 'Unmount':
                this._closeContentPanel();
                break;
            // Property pane should toggle i.e., if it's open, close it else open it.
            case 'Toggle':
                if (this.isOpen()) {
                    this.requestAction(toBeConfiguredConsumerId, 'Close', renderedByWebPart, context);
                }
                else {
                    this.requestAction(toBeConfiguredConsumerId, 'Open', renderedByWebPart, context);
                }
                break;
            // Keep the property pane open if it's already open, else no-op.
            case 'Default':
                if (this.isOpen()) {
                    // This check is to make sure that the propertypane does not obtain and render a host that is rendered
                    // in configuration tools
                    this.requestAction(toBeConfiguredConsumerId, 'Open', renderedByWebPart, context);
                }
                break;
            /**
             * Refresh the contents of the property pane if the property pane is open for the requesting component.
             *  - This condition is prechecked in 'requestPropertyPaneAction' method.
             */
            case 'Refresh':
                if (this.isOpen() && this.currentlyConfiguredConsumerId === toBeConfiguredConsumerId) {
                    this._updateContentPanelData(toBeConfiguredConsumerId).catch(function (error) {
                        sp_diagnostics_1._TraceLogger.logError(logSource_1.logSource, error);
                    });
                }
                break;
            /**
             * Refresh the contents of the property pane in the background if the property pane is not open
             * when the currently configured consumer is changed.
             */
            case 'BackgroundRefresh':
                if (!this.isOpen()) {
                    if (this.isComponentSupported(toBeConfiguredConsumerId)) {
                        sp_diagnostics_1._TraceLogger.logVerbose(logSource_1.logSource, "BackgroundRefreshing content panel");
                        this._currentlyConfiguredConsumerId = toBeConfiguredConsumerId;
                        this._updateContentPanelData(toBeConfiguredConsumerId).catch(function (error) {
                            sp_diagnostics_1._TraceLogger.logError(logSource_1.logSource, error);
                        });
                    }
                    else {
                        if (!this.isConsumerRegistered(toBeConfiguredConsumerId)) {
                            sp_diagnostics_1._TraceLogger.logVerbose(logSource_1.logSource, "Waiting to register consumer before BackgroundRefresh");
                            this._unregistredBackgroundRefreshConsumerId = toBeConfiguredConsumerId;
                        }
                        this._resetCurrentlyConfiguredConsumer();
                    }
                }
                break;
        }
    };
    ContentPanelController.prototype.isRenderedByConsumer = function () {
        return !!this._isPaneRenderedByWebPart;
    };
    /**
     * Determines if configuration tools or page details is open in the content panel.
     *
     * Warning: This returns false if other panels are open.
     */
    ContentPanelController.prototype.isOpen = function () {
        var stateContext = sp_pages_content_panel_context_1.SPPagesContentPanelProviderContext.getInstance().stateContext;
        return ((stateContext === null || stateContext === void 0 ? void 0 : stateContext.activePivot) === sp_pages_content_panel_context_1.SPPagesContentPanelPath.configurationTools && this.isAnyContentPanelOpen());
    };
    ContentPanelController.prototype.isAnyContentPanelOpen = function () {
        var _a;
        return !!((_a = sp_pages_content_panel_context_1.SPPagesContentPanelProviderContext.getInstance().stateContext) === null || _a === void 0 ? void 0 : _a.isVisible);
    };
    ContentPanelController.prototype.forceClose = function () {
        var _a;
        // let close logic run before we setIsVisible false
        this._closeContentPanel(!(0, KillSwitches_1.isForceCloseWithoutFiringEventsKSActivated)());
        (_a = sp_pages_content_panel_context_1.SPPagesContentPanelProviderContext.getInstance().subscriberContext) === null || _a === void 0 ? void 0 : _a.setIsVisible(false);
        this._resetCurrentlyConfiguredConsumer();
        sp_diagnostics_1._TraceLogger.logVerbose(logSource_1.logSource, "Force closed the content panel");
    };
    ContentPanelController.prototype.onConsumerDelete = function (id) {
        sp_core_library_1.Validate.isNonemptyString(id, 'id');
        if (id === this.currentlyConfiguredConsumerId) {
            this._resetCurrentlyConfiguredConsumer();
        }
        // Remove consumer from map on dispose to prevent memory leak.
        this._consumers.delete(id);
    };
    ContentPanelController.prototype._setAppPagePropertyPaneTopData = function (topControl, title) {
        this._title = title;
        this._topControl = topControl;
    };
    ContentPanelController.prototype._setPropertyPaneToNarrowRender = function (narrow) {
        this._isNarrow = narrow;
    };
    /**
     * Determines if the specified component can be rendered in the Content Panel
     */
    ContentPanelController.prototype.isComponentSupported = function (instanceId, consumer) {
        if (consumer === void 0) { consumer = this._consumers.get(instanceId); }
        // guard against excluded consumers, and feature flags
        if (!consumer || !ContentPanelController.isContentPanelSupported() || excludedPanels_1.excludedPanels.has(instanceId)) {
            sp_diagnostics_1._TraceLogger.logVerbose(logSource_1.logSource, "".concat(instanceId, " is not supported. (isContentPanelSupported: ").concat(!!ContentPanelController.isContentPanelSupported(), ")"));
            return false;
        }
        // already registered consumers and non webparts that are not excluded are supported
        if (this.isConsumerRegistered(instanceId) || !(0, propertyPaneConsumer_1.isConsumerWithManifest)(consumer) || !(0, propertyPaneConsumer_1.isWebpart)(consumer)) {
            return true;
        }
        var verifiedManifest = sp_loader_1.SPComponentLoader.tryGetManifestById(consumer.context.manifest.id || '');
        // webparts
        return !!(typeof (verifiedManifest === null || verifiedManifest === void 0 ? void 0 : verifiedManifest.isInternal) === 'boolean'
            ? verifiedManifest.isInternal
            : (verifiedManifest === null || verifiedManifest === void 0 ? void 0 : verifiedManifest.id) && includedWebpart_1.includedWebparts.has(verifiedManifest.id));
    };
    /**
     * Callback to handle the configuration events that originate in the property pane.
     *
     * @example
     * PropertyPaneClosed, PropertyPaneApplyClicked etc.,
     */
    ContentPanelController.prototype._onConfigurationEvent = function (configurationEvent) {
        sp_diagnostics_1._TraceLogger.logVerbose(logSource_1.logSource, "Configuration event '".concat(configurationEvent, "'. (currentlyConfiguredConsumerId: ").concat(this._currentlyConfiguredConsumerId, ")"));
        switch (configurationEvent) {
            /**
             * For the event 'PropertyPaneClosed' we need to fire both the 'PropertyPaneClosed' and
             * 'PropertyPaneConfigurationComplete' events to the consumer.
             * '_togglePropertyPanePosition' internally fires both the events.
             */
            case 'Closed':
                this.requestAction(this.currentlyConfiguredConsumerId || '', 'Close', this.isRenderedByConsumer());
                break;
            /**
             * For these events we need to fire two events:
             *  - 'PropertyPaneConfigurationComplete' and
             *  - Whatever the event the caller has passed in 'configurationEvent'.
             */
            case 'ApplyClicked':
            case 'LostFocus':
                this._fireConfigurationEvent(this._currentlyConfiguredConsumerId, configurationEvent);
                this._fireConfigurationEvent(this._currentlyConfiguredConsumerId, 'ConfigurationComplete');
                break;
            case 'BackClicked':
                !(0, KillSwitches_1.isShowBackButtonBackKSActivated)() &&
                    this._fireConfigurationEvent(this._currentlyConfiguredConsumerId, configurationEvent);
        }
    };
    /**
     * Property pane field change event handler.
     *
     * @param propertyName - Name of the property pane field changed.
     * @param newValue - New value.
     *  This value could be undefined/empty in the case of custom field.
     */
    ContentPanelController.prototype._onPropertyPaneFieldChanged = function (propertyName, newValue, fieldType) {
        sp_diagnostics_1._TraceLogger.logVerbose(logSource_1.logSource, "property pane field '".concat(propertyName, "' changed. currentlyConfiguredConsumerId: ").concat(this._currentlyConfiguredConsumerId));
        if (!this._currentlyConfiguredConsumerId) {
            var error = new Error('onPropertyPaneFieldChanged event is not expected when no consumer is being configured');
            sp_diagnostics_1._TraceLogger.logError(logSource_1.logSource, error);
            throw error;
        }
        var consumer = this._consumers.get(this._currentlyConfiguredConsumerId);
        if (consumer) {
            consumer._onPropertyPaneFieldChanged(propertyName, newValue, fieldType);
            // Resetting the boolean value.
            this._updateContentPanelData(this.currentlyConfiguredConsumerId || '').catch(function (error) {
                sp_diagnostics_1._TraceLogger.logError(logSource_1.logSource, error);
            });
        }
        if (consumer === null || consumer === void 0 ? void 0 : consumer._isPropertyPaneReactive()) {
            this._resetConfigurationCompleteTimeout();
        }
    };
    /**
     * Method to fire the configuration event to the host and the consumer.
     *
     * @param componentId - Id of the consumer to which the event needs to be sent.
     * @param configurationEvent - Kind of configuration event to fire.
     */
    ContentPanelController.prototype._fireConfigurationEvent = function (componentId, configurationEvent) {
        this._clearConfigurationCompleteTimeout();
        // Up to IWebPartGetter to handle an undefined string.
        var consumer = this._consumers.get(componentId || '');
        if (consumer) {
            var eventArgs = {
                componentId: componentId,
                configurationEvent: configurationEvent
            };
            sp_core_library_1._SPEventManager.instance.raiseEvent('propertyPaneEvent', eventArgs);
            // Raise the event to the consumer.
            consumer._onPropertyPaneLifeCycleEvent(configurationEvent);
        }
    };
    /**
     * Reset configuration completion timeout.
     */
    ContentPanelController.prototype._resetConfigurationCompleteTimeout = function () {
        this._clearConfigurationCompleteTimeout();
        this._configurationCompletionTimeout = window.setTimeout(this._fireConfigurationEvent, constants_1.CONFIGURATION_COMPLETE_TIMEOUT, this._currentlyConfiguredConsumerId, 'ConfigurationComplete');
    };
    /**
     * Clear configuration completion timeout.
     */
    ContentPanelController.prototype._clearConfigurationCompleteTimeout = function () {
        if (this._configurationCompletionTimeout) {
            window.clearTimeout(this._configurationCompletionTimeout);
            this._configurationCompletionTimeout = undefined;
        }
    };
    /**
     * Opens the supported panel in the content panel for the specified componentId.
     * Content Panel may support different consumer panels and this will handle the logic for each consumer
     * and render them in their respective panel.
     */
    ContentPanelController.prototype._openContentPanel = function (componentId, renderedByWebPart) {
        var args = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            args[_i - 2] = arguments[_i];
        }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var paneOpenBeforeRender, newConsumer, subscriberContext;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.isConsumerRegistered(componentId)) {
                            return [2 /*return*/];
                        }
                        paneOpenBeforeRender = !!this.isOpen();
                        newConsumer = this._consumers.get(componentId);
                        if (componentId !== this._currentlyConfiguredConsumerId) {
                            /**
                             * Premptively asking consumer(if exists, consumer can be undefined for canvas toolbox hint.)
                             * to load async resources for its property pane before the data
                             * is requested via _getPropertyPaneData call.
                             * We make this call when the active webpart has changed and the property pane is shown.
                             */
                            newConsumer === null || newConsumer === void 0 ? void 0 : newConsumer._loadPropertyPaneResources();
                            // Raise the event to the consumer.
                            this._fireConfigurationEvent(componentId, 'ActiveWebPartChanged');
                        }
                        /**
                         * If pane is 'open' before '_renderPropertyPane' is called that means property pane did not toggle.
                         * Hence no events were fired. So fire the events explicitly.
                         */
                        if (paneOpenBeforeRender) {
                            // Firing events for the current consumer.
                            this._fireConfigurationEvent(this._currentlyConfiguredConsumerId, 'ConfigurationComplete');
                            // Firing events for the new consumer.
                            this._fireConfigurationEvent(componentId, 'ConfigurationStart');
                        }
                        else {
                            // Raise the event to the consumer.
                            this._fireConfigurationEvent(componentId, 'ConfigurationStart');
                            this._fireConfigurationEvent(componentId, 'Opened');
                        }
                        this._isPaneRenderedByWebPart = renderedByWebPart;
                        this._currentlyConfiguredConsumerId = componentId;
                        return [4 /*yield*/, sp_pages_content_panel_context_1.SPPagesContentPanelProviderContext.getInstance().waitForContext];
                    case 1:
                        subscriberContext = (_a.sent()).subscriberContext;
                        subscriberContext.setIsVisible(sp_pages_content_panel_context_1.SPPagesContentPanelPath.configurationTools);
                        sp_diagnostics_1._TraceLogger.logVerbose(logSource_1.logSource, "Open panel for '".concat(componentId, "' (renderedByWebPart: ").concat(renderedByWebPart, ")"));
                        return [4 /*yield*/, this._updateContentPanelData.apply(this, tslib_1.__spreadArray([componentId], args, false))];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Function that updates the data in the content panel
     */
    ContentPanelController.prototype._updateContentPanelData = function (componentId) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var subscriberContext;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        sp_diagnostics_1._TraceLogger.logVerbose(logSource_1.logSource, "Update property pane data for ".concat(componentId));
                        return [4 /*yield*/, sp_pages_content_panel_context_1.SPPagesContentPanelProviderContext.getInstance().waitForContext];
                    case 1:
                        subscriberContext = (_a.sent()).subscriberContext;
                        subscriberContext.setLoader(sp_pages_content_panel_context_1.SPPagesContentPanelPath.configurationTools, function () {
                            var _a;
                            var qosMonitor = new sp_diagnostics_1._QosMonitor((0, sp_pages_content_panel_context_1.getQOSName)(logSource_1.COMPONENT_NAME, sp_pages_content_panel_context_1.SPPagesContentPanelQOSScenarios.rendering));
                            return (_a = _this._consumers
                                .get(componentId)) === null || _a === void 0 ? void 0 : _a._getPropertyPaneData.apply(_a, args).then(function (propertyPaneData) {
                                sp_diagnostics_1._TraceLogger.logVerbose(logSource_1.logSource, "Property pane data returned for ".concat(componentId));
                                qosMonitor.writeSuccess();
                                return {
                                    propertyPaneData: tslib_1.__assign(tslib_1.__assign({}, propertyPaneData), { title: _this._title || propertyPaneData.title }),
                                    componentId: componentId,
                                    isNarrow: _this._isNarrow,
                                    topControl: _this._topControl
                                };
                            }).catch(function (error) {
                                var propertyPaneConsumer = _this._consumers.get(componentId);
                                var customError = new odsp_core_bundle_1.CustomError({
                                    message: (0, odsp_core_bundle_1.markPrivacyData)("Failed to get property pane data for ".concat((0, propertyPaneConsumer_1.isConsumerWithManifest)(propertyPaneConsumer)
                                        ? propertyPaneConsumer.context.manifest.id
                                        : componentId), 'safe'),
                                    innerError: error
                                });
                                sp_diagnostics_1._TraceLogger.logError(logSource_1.logSource, customError);
                                qosMonitor.writeUnexpectedFailure('UnhandledError', error);
                                return undefined;
                            });
                        });
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Resets the currently configured consumer and resets the data hosted
     * in the respective panel's.
     *
     * Note this does not close the content panel
     */
    ContentPanelController.prototype._resetCurrentlyConfiguredConsumer = function () {
        var _a;
        sp_diagnostics_1._TraceLogger.logVerbose(logSource_1.logSource, "Removing data for content panel");
        this._currentlyConfiguredConsumerId = undefined;
        this._isPaneRenderedByWebPart = undefined;
        (_a = sp_pages_content_panel_context_1.SPPagesContentPanelProviderContext.getInstance().subscriberContext) === null || _a === void 0 ? void 0 : _a.setLoader(sp_pages_content_panel_context_1.SPPagesContentPanelPath.configurationTools, function () { return undefined; });
    };
    /**
     * Closes the content panel if the current open panel is a panel supported by this component.
     */
    ContentPanelController.prototype._closeContentPanel = function (skipFiringEvents) {
        var _a = sp_pages_content_panel_context_1.SPPagesContentPanelProviderContext.getInstance(), stateContext = _a.stateContext, subscriberContext = _a.subscriberContext;
        if ((stateContext === null || stateContext === void 0 ? void 0 : stateContext.isVisible) && (stateContext === null || stateContext === void 0 ? void 0 : stateContext.activePivot) === sp_pages_content_panel_context_1.SPPagesContentPanelPath.configurationTools) {
            subscriberContext === null || subscriberContext === void 0 ? void 0 : subscriberContext.setIsVisible(false);
            if (!skipFiringEvents) {
                this._fireConfigurationEvent(this._currentlyConfiguredConsumerId, 'ConfigurationComplete');
                this._fireConfigurationEvent(this._currentlyConfiguredConsumerId, 'Closed');
            }
            sp_diagnostics_1._TraceLogger.logVerbose(logSource_1.logSource, "Closed content panel");
        }
    };
    return ContentPanelController;
}());
exports.ContentPanelController = ContentPanelController;
