// @microsoft/sp-property-pane/lib-commonjs/propertyPaneFields/propertyPaneSortableAccordion/PropertyPaneSortableAccordion.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useStyles = void 0;
var tslib_1 = require("tslib");
var react_components_1 = require("@fluentui/react-components");
exports.useStyles = (0, react_components_1.makeStyles)({
    accordionItem: {
        display: 'flex',
        flexDirection: 'row',
        alignItems: 'center',
        position: 'relative',
        '&:hover .deleteButton, &:focus-within .deleteButton, &:focus-visible .deleteButton': {
            display: 'flex'
        },
        '::before': {
            content: '""',
            display: 'var(--focusIndicatorDisplay)',
            position: 'absolute',
            backgroundColor: react_components_1.tokens.colorBrandBackground,
            width: '2px',
            height: '100%',
            left: '-20px'
        }
    },
    accordionItemHeader: tslib_1.__assign(tslib_1.__assign({ display: 'flex', flexDirection: 'row', alignItems: 'center', columnGap: '8px' }, react_components_1.shorthands.padding('4px', '0px')), { width: '100%' }),
    accordionItemTitle: tslib_1.__assign(tslib_1.__assign({ flexShrink: 1, flexGrow: 1 }, react_components_1.shorthands.overflow('hidden')), { textOverflow: 'ellipsis', whiteSpace: 'nowrap', cursor: 'pointer' }),
    accordionItemButtons: {
        flexShrink: 0,
        display: 'flex',
        flexDirection: 'row',
        gap: '10px'
    },
    dragButton: tslib_1.__assign({ cursor: 'move', flexShrink: 0 }, react_components_1.shorthands.border('none')),
    deleteButton: tslib_1.__assign({ display: 'none' }, react_components_1.shorthands.border('none')),
    collapseButton: tslib_1.__assign({}, react_components_1.shorthands.border('none')),
    itemConfigurePane: tslib_1.__assign({}, react_components_1.shorthands.padding('0', '8px'))
});
