// @microsoft/sp-http-base/lib-commonjs/oauthTokenProvider/AadKillSwitches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isUseSessionStorageForRedirectScenarioProxyMonitoringKSActivated = exports.isEnableEncryptedAuthorizeResponseKSActivated = exports.isEnablePopupFlowForMsalV3KSActivated = exports.isDisableMsalV1KSActivated = exports.isOnlyEnsureServicePrincipalIdOnTokenFetchKSActivated = exports.isFailAllEmailAuthGuestTokenRequestsKSActivated = exports.isFailAllAnonymousTokenRequestsKSActivated = exports.isLogAllFullPageRedirectsKSActivated = exports.isConditionalAccessIgnoredOnEnsureStateKSActivated = exports.isUpdateReplyUriForV1FallbackKSActivated = exports.isEnsureRefreshTokenKSActivated = exports.isTenantAdminAppFirtPartyKSActivated = exports.isAlwaysInitOboConfigKSActive = exports.isMsalImplicitV1TokenByPopupKSActivated = exports.isLogAfterFullPageRedirectKSActivated = exports.useOBOInWebForSecureBrokerKSActivated = exports.isTeamsBrowserHostedKSActive = exports.isPreventInfiniteRefreshViaReadingFromSessionStorageKSActivated = exports.isMsalUseSeparateStorageKeysKsActivated = exports.isMsalBrowserRetryAndSsoSilentImprovementsKsActivated = exports.isMsalTwoMinuteMaxRedirectKsActivated = exports.isManualRedirectCancellationActivated = exports.isAadTokenErrorLogKsActivated = exports.isExtendedErrorObjectActivated = exports.isSkipKnownErroneousResourcesActivated = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
function isSkipKnownErroneousResourcesActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('dd3dba4e-4baa-4bd3-95bf-64ffae9b7b51'
    /* '01/04/22', 'Skip making a request for resources that have already failed' */
    );
}
exports.isSkipKnownErroneousResourcesActivated = isSkipKnownErroneousResourcesActivated;
function isExtendedErrorObjectActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('a4a3d2f4-ad0e-4684-adee-a5ceec193558'
    /* '03/08/22', 'Extends error object to allow for more properties to be filled' */
    );
}
exports.isExtendedErrorObjectActivated = isExtendedErrorObjectActivated;
function isAadTokenErrorLogKsActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e212149a-997e-4aa9-aa89-8be796153d3e'
    /* '05/27/2022',
        'Log actual acquire token monitor error'*/
    );
}
exports.isAadTokenErrorLogKsActivated = isAadTokenErrorLogKsActivated;
function isManualRedirectCancellationActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('3c6eb7e3-520b-41c6-b517-2f40a6dc2989'
    /* '08/16/2024',
        'bcleaver_ManuallyCancelRedirect'*/
    );
}
exports.isManualRedirectCancellationActivated = isManualRedirectCancellationActivated;
function isMsalTwoMinuteMaxRedirectKsActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('ca6e64a5-3cde-4350-8a45-a21de378cc60'
    /* '07/14/22', 'Enable MSAL to hit max redirect limit in 2 minutes instead of 1' */
    );
}
exports.isMsalTwoMinuteMaxRedirectKsActivated = isMsalTwoMinuteMaxRedirectKsActivated;
function isMsalBrowserRetryAndSsoSilentImprovementsKsActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e0bd3586-89a3-466b-8ee0-db1f9457fe99'
    /* '07/14/22', 'Stop MSAL Browser from retrying endpoint resolution failures + increase timeout time' */
    );
}
exports.isMsalBrowserRetryAndSsoSilentImprovementsKsActivated = isMsalBrowserRetryAndSsoSilentImprovementsKsActivated;
function isMsalUseSeparateStorageKeysKsActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('1a423106-78a1-4238-ba44-dd5f7f4138df'
    /* '08/18/22', 'Separate the storage parameters used for redirect flows by v1/v2' */
    );
}
exports.isMsalUseSeparateStorageKeysKsActivated = isMsalUseSeparateStorageKeysKsActivated;
function isPreventInfiniteRefreshViaReadingFromSessionStorageKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('fed6018f-7453-47fc-aecb-2d9505b62bb4'
    /* '03/02/23', 'Add session storage check to prevent infinite refresh' */
    );
}
exports.isPreventInfiniteRefreshViaReadingFromSessionStorageKSActivated = isPreventInfiniteRefreshViaReadingFromSessionStorageKSActivated;
function isTeamsBrowserHostedKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('9b9957d5-67d6-4f73-8882-11ee6bbaa066'
    /* '02/20/2024', 'Set teams hosted container' */
    );
}
exports.isTeamsBrowserHostedKSActive = isTeamsBrowserHostedKSActive;
function useOBOInWebForSecureBrokerKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('043feae9-5a56-40ff-b175-6f7f813507ff' /* '10/26/22', 'use OBO in web for secure broker' */);
}
exports.useOBOInWebForSecureBrokerKSActivated = useOBOInWebForSecureBrokerKSActivated;
function isLogAfterFullPageRedirectKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('5a350491-8d32-46a3-a4c4-fdce19a5f73e' /* '08/17/23', 'log after full page redirect' */);
}
exports.isLogAfterFullPageRedirectKSActivated = isLogAfterFullPageRedirectKSActivated;
function isMsalImplicitV1TokenByPopupKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e6246a29-f6af-41de-88b7-3f023102181b' /* '08/14/23', 'Enable acquiring a token by popup' */);
}
exports.isMsalImplicitV1TokenByPopupKSActivated = isMsalImplicitV1TokenByPopupKSActivated;
// cross project KS
function isAlwaysInitOboConfigKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('7fb9acc4-19f2-446f-b592-f15c49bfd771'
    /* '10/17/2023', 'Always initialize OBO configuration' */
    );
}
exports.isAlwaysInitOboConfigKSActive = isAlwaysInitOboConfigKSActive;
function isTenantAdminAppFirtPartyKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('43551625-af3d-44b0-8fb3-f0272435c606' /* '12/08/23', 'Is tenant admin app first party' */);
}
exports.isTenantAdminAppFirtPartyKSActivated = isTenantAdminAppFirtPartyKSActivated;
function isEnsureRefreshTokenKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('db350345-b6f8-4c00-9cbe-49ae316f6637' /* '12/06/2023', 'Disable ensure refresh token' */);
}
exports.isEnsureRefreshTokenKSActivated = isEnsureRefreshTokenKSActivated;
function isUpdateReplyUriForV1FallbackKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('bd15098f-e57e-4171-8d51-c5c0c59dfaeb' /* '01/31/2024', 'isUpdateReplyUriForV1Fallback' */);
}
exports.isUpdateReplyUriForV1FallbackKSActivated = isUpdateReplyUriForV1FallbackKSActivated;
function isConditionalAccessIgnoredOnEnsureStateKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('780272c0-bb66-4f68-a470-15c7c74493fd' /* '02/09/2024', 'isConditionalAccessIgnoredOnEnsureStateKSActivated' */);
}
exports.isConditionalAccessIgnoredOnEnsureStateKSActivated = isConditionalAccessIgnoredOnEnsureStateKSActivated;
function isLogAllFullPageRedirectsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('33e01c38-ae10-4a40-ae92-1d8f82214515' /* '04/24/2024', 'isLogAllFullPageRedirectsKSActivated' */);
}
exports.isLogAllFullPageRedirectsKSActivated = isLogAllFullPageRedirectsKSActivated;
function isFailAllAnonymousTokenRequestsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('bba4380d-7882-4c9f-9033-0dba0b76ca93' /* '05/30/2024', 'isFailAllAnonymousTokenRequestsKSActivated' */);
}
exports.isFailAllAnonymousTokenRequestsKSActivated = isFailAllAnonymousTokenRequestsKSActivated;
function isFailAllEmailAuthGuestTokenRequestsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('52e15dc9-ebf3-41a7-af0f-ff9cf93ac3bd' /* '05/30/2024', 'isFailAllEmailAuthGuestTokenRequestsKSActivated' */);
}
exports.isFailAllEmailAuthGuestTokenRequestsKSActivated = isFailAllEmailAuthGuestTokenRequestsKSActivated;
function isOnlyEnsureServicePrincipalIdOnTokenFetchKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('748cee93-1dd8-4510-86b8-56446240c126' /* '05/13/2024', 'isOnlyEnsureServicePrincipalIdOnTokenFetchKSActivated' */);
}
exports.isOnlyEnsureServicePrincipalIdOnTokenFetchKSActivated = isOnlyEnsureServicePrincipalIdOnTokenFetchKSActivated;
function isDisableMsalV1KSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('f2a68e00-be7a-4610-be2c-5f6ef5cfe42b' /* '07/08/2024', 'isDisableMsalV1KSActivated' */);
}
exports.isDisableMsalV1KSActivated = isDisableMsalV1KSActivated;
function isEnablePopupFlowForMsalV3KSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('32a6db75-625e-451a-8813-6a88e019e71f' /* '07/01/2024', 'isPopupFlowEnabledForMsalV3' */);
}
exports.isEnablePopupFlowForMsalV3KSActivated = isEnablePopupFlowForMsalV3KSActivated;
function isEnableEncryptedAuthorizeResponseKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e2de243c-2222-4297-be12-92b853b1a90d' /* '07/23/2024', 'isEnableEncryptedAuthorizeResponseKSActivated' */);
}
exports.isEnableEncryptedAuthorizeResponseKSActivated = isEnableEncryptedAuthorizeResponseKSActivated;
function isUseSessionStorageForRedirectScenarioProxyMonitoringKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('57d6fc0d-c2df-4446-8a94-d9ccc9e8870b' /* '08/13/2024', 'isUseSessionStorageForRedirectScenarioProxyMonitoringKSActivated' */);
}
exports.isUseSessionStorageForRedirectScenarioProxyMonitoringKSActivated = isUseSessionStorageForRedirectScenarioProxyMonitoringKSActivated;
