// @microsoft/sp-loader/lib-commonjs/tiny/common/flights.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isUseFirstPartyForThirdPartyAuthenticationFlightEnabled = exports.isNaaFlightEnabled = exports.isSpfxSubResourceIntegrityFlightEnabled = exports.isMsalBrowserV3FlightEnabled = exports.isAPEXSPLoaderRemoveSystemJSEnabled = exports.isAPEXSpfxCustomAppPrincipalClientEnabled = exports.isOBOForAllTeamsScenariosEnabled = exports.isMinimalStartFlightEnabled = exports.isSafariAuthPatchEnabled = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
/**
 * SppplatSafariAuthPatch
 */
function isSafariAuthPatchEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(687);
}
exports.isSafariAuthPatchEnabled = isSafariAuthPatchEnabled;
/**
 * SppplatMinimalStart
 */
function isMinimalStartFlightEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(1685);
}
exports.isMinimalStartFlightEnabled = isMinimalStartFlightEnabled;
/**
 * APEXUseOboForAllTeamsScenarios
 */
function isOBOForAllTeamsScenariosEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(60244);
}
exports.isOBOForAllTeamsScenariosEnabled = isOBOForAllTeamsScenariosEnabled;
/**
 * APEXSpfxCustomAppPrincipalClient
 * https://spoflights.azurewebsites.net/Flight/Edit/APEXSpfxCustomAppPrincipalClient
 */
function isAPEXSpfxCustomAppPrincipalClientEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(1807);
}
exports.isAPEXSpfxCustomAppPrincipalClientEnabled = isAPEXSpfxCustomAppPrincipalClientEnabled;
/**
 * APEXSPLoaderRemoveSystemJS
 * https://spoflights.azurewebsites.net/Flight/Edit/APEXSPLoaderRemoveSystemJS
 */
function isAPEXSPLoaderRemoveSystemJSEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(60360);
}
exports.isAPEXSPLoaderRemoveSystemJSEnabled = isAPEXSPLoaderRemoveSystemJSEnabled;
/**
 * MsalJsV3Migration
 * https://spoflights.azurewebsites.net/Flight/Edit/MsalJsV3Migration
 */
function isMsalBrowserV3FlightEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(61223);
}
exports.isMsalBrowserV3FlightEnabled = isMsalBrowserV3FlightEnabled;
/**
 * SpfxSubresourceIntegrity
 * https://spoflights.azurewebsites.net/Flight/Edit/SpfxSubresourceIntegrity
 */
function isSpfxSubResourceIntegrityFlightEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(60737);
}
exports.isSpfxSubResourceIntegrityFlightEnabled = isSpfxSubResourceIntegrityFlightEnabled;
/**
 * APEXTeamsNaaEnabled
 * https://spoflights.azurewebsites.net/Flight/Edit/APEXTeamsNaaEnabled
 */
function isNaaFlightEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(61858);
}
exports.isNaaFlightEnabled = isNaaFlightEnabled;
/**
 * APEXUseFirstPartyForThirdPartyAuthentication
 * https://spoflights.azurewebsites.net/Flight/Edit/APEXUseFirstPartyForThirdPartyAuthentication
 */
function isUseFirstPartyForThirdPartyAuthenticationFlightEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(61963);
}
exports.isUseFirstPartyForThirdPartyAuthenticationFlightEnabled = isUseFirstPartyForThirdPartyAuthenticationFlightEnabled;
